File: backend/backend_crypto_purchase_history/src/services/transactionsService.js
const { query } = require("../db");

class TransactionsError extends Error {
  constructor(message, statusCode) {
    super(message);
    this.name = "TransactionsError";
    this.statusCode = statusCode;
  }
}

function parsePagination(page, limit) {
  const parsedPage = Number(page || 1);
  const parsedLimit = Number(limit || 20);

  if (!Number.isInteger(parsedPage) || parsedPage < 1) { throw new TransactionsError("page must be a positive integer", 400); }
  if (!Number.isInteger(parsedLimit) || parsedLimit < 1 || parsedLimit > 100) { throw new TransactionsError("limit must be an integer between 1 and 100", 400); }

  return {
    page: parsedPage,
    limit: parsedLimit,
    offset: (parsedPage - 1) * parsedLimit
  };
}

function buildTransactionsService(deps = {}) {
  const runQuery = deps.query || query;

  return {
    async getUserTransactions({ userId, page, limit }) {
      if (!Number.isInteger(userId) || userId <= 0) { throw new TransactionsError("Authenticated user is required", 401); }
      const { page: resolvedPage, limit: resolvedLimit, offset } = parsePagination(page, limit);
      const countRows = await runQuery(
        "SELECT COUNT(*) AS total FROM transactions WHERE user_id = ?",
        [userId]
      );
      const total = Number(countRows[0]?.total || 0);
      const rows = await runQuery(
        `SELECT transaction_type, quantity, price_usd, created_at
        FROM transactions
        WHERE user_id = ?
        ORDER BY created_at DESC, id DESC
        LIMIT ? OFFSET ?`,
        [userId, resolvedLimit, offset]
      );
      const transactions = rows.map((row) => ({
        type: row.transaction_type,
        amount: row.quantity,
        priceAtTime: row.price_usd,
        timestamp: row.created_at
      }));

      return {
        data: transactions,
        pagination: {
          page: resolvedPage,
          limit: resolvedLimit,
          total,
          totalPages: total === 0 ? 0 : Math.ceil(total / resolvedLimit)
        }
      };
    }
  };
}

module.exports = {
  TransactionsError,
  buildTransactionsService
};

File: backend/backend_Crypto_Purchase_History/src/main/java/com/cs506/cryptohistory/service/TransactionsService.java
package com.cs506.cryptohistory.service;

import com.cs506.cryptohistory.dto.PaginationDto;
import com.cs506.cryptohistory.dto.TransactionHistoryItem;
import com.cs506.cryptohistory.dto.TransactionHistoryResponse;
import com.cs506.cryptohistory.exception.TransactionHistoryException;
import com.cs506.cryptohistory.model.TransactionRow;
import com.cs506.cryptohistory.repository.TransactionHistoryRepository;
import java.util.List;
import org.springframework.stereotype.Service;

@Service
public class TransactionsService {

    private final TransactionHistoryRepository repository;

    public TransactionsService(TransactionHistoryRepository repository) {
        this.repository = repository;
    }

    public TransactionHistoryResponse getUserTransactions(Integer userId, Integer page, Integer limit) {
        if (userId == null || userId <= 0) {
            throw new TransactionHistoryException("Authenticated user is required", 401);
        }

        int resolvedPage = page == null ? 1 : page;
        int resolvedLimit = limit == null ? 20 : limit;

        if (resolvedPage < 1) {
            throw new TransactionHistoryException("page must be a positive integer", 400);
        }
        if (resolvedLimit < 1 || resolvedLimit > 100) {
            throw new TransactionHistoryException("limit must be an integer between 1 and 100", 400);
        }

        int offset = (resolvedPage - 1) * resolvedLimit;
        int total = repository.countByUserId(userId);
        List<TransactionRow> rows = repository.findByUserId(userId, resolvedLimit, offset);
        List<TransactionHistoryItem> items = rows.stream()
            .map(row -> new TransactionHistoryItem(
                row.transactionType(),
                row.quantity(),
                row.priceUsd(),
                row.createdAt()
            ))
            .toList();

        int totalPages = total == 0 ? 0 : (int) Math.ceil((double) total / resolvedLimit);
        return new TransactionHistoryResponse(
            items,
            new PaginationDto(resolvedPage, resolvedLimit, total, totalPages)
        );
    }
}
